### PR TITLE
Provide context when `?` can't be called because of `Result<_, E>`

### LINKED ISSUE #72124: Rustc fails to suggest an easy fix; instead provides a confusing error message
The following code produces an unexpected error message:

```rust
test.split_whitespace().next().ok_or_else(|| {
    format_err!("Couldn't split the test string");
})?;
```

I expected the compiler to give me a warning like "`ok_or_else` needs to return the trait `std::error::Error`. Instead you are returning `()`. Suggestion: remove the `;`.

Instead, I received the following error message which was hard to parse:
```
 |     })?;
  |       ^ the trait `std::error::Error` is not implemented for `()`
  |
  = note: required because of the requirements on the impl of `failure::Fail` for `()`
  = note: required because of the requirements on the impl of `std::convert::From<()>` for `failure::Error`
  = note: required by `std::convert::From::from`
```

The rust compiler version that I am using is `1.43.1`. This behavior also exists within the beta and nightly versions. I also created an example using the rust-playground here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=33e42842b0d6385ea5704e323c113e7f.

Thanks again!

### PR DESCRIPTION
**Pull request by estebank**
When a method chain ending in `?` causes an E0277 because the expression's `Result::Err` variant doesn't have a type that can be converted to the `Result<_, E>` type parameter in the return type, provide additional context of which parts of the chain can and can't support the `?` operator.

```
error[E0277]: `?` couldn't convert the error to `String`
  --> $DIR/question-mark-result-err-mismatch.rs:27:25
   |
LL | fn bar() -> Result<(), String> {
   |             ------------------ expected `String` because of this
LL |     let x = foo();
   |             ----- this has type `Result<_, String>`
...
LL |         .map_err(|_| ())?;
   |          ---------------^ the trait `From<()>` is not implemented for `String`
   |          |
   |          this can't be annotated with `?` because it has type `Result<_, ()>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the following other types implement trait `From<T>`:
             <String as From<char>>
             <String as From<Box<str>>>
             <String as From<Cow<'a, str>>>
             <String as From<&str>>
             <String as From<&mut str>>
             <String as From<&String>>
   = note: required for `Result<(), String>` to implement `FromResidual<Result<Infallible, ()>>`
```

Fix #72124.

### CODE REVIEW

File: tests/ui/traits/question-mark-result-err-mismatch.stderr [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+   | |__________- this can be annotated with `?` because it has type `Result<&str, &str>`
+LL |       let one = x
+LL |           .map(|s| ())
+   |            ----------- this can be annotated with `?` because it has type `Result<(), &str>`
+LL |           .map_err(|_| ())
+   |            --------------- this can't be annotated with `?` because it has type `Result<(), ()>`
+LL |           .map(|()| "")?;
+   |            ------------^ the trait `From<()>` is not implemented for `String`
+   |            |
+   |            this can't be annotated with `?` because it has type `Result<&str, ()>`
```
Comment (compiler-errors) [inline tests/ui/traits/question-mark-result-err-mismatch.stderr:20]: this seems really heavy for long chains -- do i really care about `?`-ability 4 `.map`s down?

**Comment by compiler-errors** [comment]
Also, this doesn't really fix #72124, does it? The problem in that issue is a shortcoming of closure inference causing us to forgo suggesting removing `;` in a closure body?

File: tests/ui/traits/question-mark-result-err-mismatch.stderr (hunk 1 above)
Comment (estebank) [reply tests/ui/traits/question-mark-result-err-mismatch.stderr:20]: Probably not. I would like to add an extra check to only point at places where the `E` *changed* (like we do for assoc types in method call chains), but wanted to keep the code down to a reviewable size.

Comment (compiler-errors) [reply tests/ui/traits/question-mark-result-err-mismatch.stderr:20]: I'd rather just put a hard limit on the number of method chain to walk backwards. Maybe 2 is sensible.

**Comment by estebank** [comment]
> Also, this doesn't really fix https://github.com/rust-lang/rust/issues/72124, does it?

I'm adding some checks for the specific methods that take closures and that can change the type of `Result<_, E>`, but we still need the general context for any other cases that have underlying reasons beyond `map_err` calls.

**Comment by compiler-errors** [comment]
> I'm adding some checks for the specific methods that take closures 

For what it's worth, I think those checks should be implemented in a separate PR to separate concerns. 

I'm mostly just noting that the PR shouldn't close (via `fixes ###` in the PR summary) that issue since there's additional work that needs to be done on that case specifically.

File: tests/ui/traits/question-mark-result-err-mismatch.stderr (hunk 1 above)
Comment (estebank) [reply tests/ui/traits/question-mark-result-err-mismatch.stderr]: Let me know what you think of the new output.

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs [hunk 2]
```diff
@@ -945,6 +957,223 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
+                        if let hir::ExprKind::Call(_, [expr, ..]) = expr.kind {
+                            self.found = Some(expr);
+                            return;
+                        }
+                    }
+                }
+                hir::intravisit::walk_expr(self, ex);
+            }
+        }
+        let hir_id = self.tcx.hir().local_def_id_to_hir_id(obligation.cause.body_id);
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:988]: This needs a rebase.

**Review by compiler-errors — CHANGES_REQUESTED** [review]
please amend commit 2faaea98f813827ca0f91266652039d2d859989d after rebasing because `local_def_id_to_hir_id` has moved onto `TyCtxt`.

**Comment by compiler-errors** [comment]
@bors r-

r=me after fixing up that first commit

**Comment by compiler-errors** [comment]
(also if you're already changing this, pls squash the formatting commits down into more relevant ones)

**Comment by compiler-errors** [comment]
Thanks!

@bors r+
